### ISSUE TITLE
EnumSerializer fails when when multiple enums are used

### ISSUE THREAD

**Issue by aboisvert** [lift/framework#1080]
lift-json-ext's EnumSerializer and EnumNameSerializer do not work properly when multiple enums are used to serialize/deserialize data.

The root cause appears to be the use of,

<pre><code>  val EnumerationClass = classOf[E#Value]
</code></pre>

which returns the same class for all Enumeration-deriving classes, and therefore the subsequent match:

<pre><code>    PartialFunction[(TypeInfo, JValue), E#Value] = {
      case (TypeInfo(EnumerationClass, _), json) => json match {
</code></pre>

matches against any possible enums, leading to -- at best -- a runtime failure because the enum value cannot be found, or -- at worst -- a sneaky failure since a different enum value may be returned.

I'm attaching a test that demonstrates the issue with EnumNameSerializer.

**Comment by fbettag** [comment]
Hey there, this can be solved with existing tools like mentioned in [this blogpost here](http://blog.knoldus.com/2012/05/18/liftweb-json-serialization-and-deserialization-using-scala/).
By having a single Enum to match this against, this apparently works.

```scala
/*
*  Serializer and Deserializers for Enumeration
*/
class EnumerationSerializer(enumList: List[Enumeration]) extends net.liftweb.json.Serializer[Enumeration#Value] {
  import JsonDSL._
  val EnumerationClass = classOf[Enumeration#Value]
  val formats = Serialization.formats(NoTypeHints)
 
 //  Deserializer Function for Enumeration
  def deserialize(implicit format: Formats): PartialFunction[(TypeInfo, JValue), Enumeration#Value] = { 
    case (TypeInfo(EnumerationClass, _), json) => json match {
      case JObject(List(JField(name, JString(value)))) => fetchEnumValue(enumList, value)  // Here we can perform the
      case JString(value) => fetchEnumValue(enumList, value)                               // desired function
      case value => throw new MappingException("Can't convert " +
        value + " to " + EnumerationClass)
    }
  }
 
  def serialize(implicit format: Formats): PartialFunction[Any, JValue] = {
    case i: Enumeration#Value => i.toString
  }
 
  private def fetchEnumValue(enumList: List[Enumeration], value: String): Enumeration#Value = {
    var defaultEnumValue: Enumeration#Value = Year.withName("FinalYear")
    for (enumItem <- enumList) {
      for (enumValue <- enumItem.values) {
        enumValue.toString == value match {
          case true => {
            defaultEnumValue = enumItem.withName(value)
          }
          case _ => None
        }
      }
    }
    defaultEnumValue
  }
 
}
```

**Comment by Shadowfiend** [comment]
Yes, I believe that's very similar to the solution [holograph posted above](https://github.com/lift/framework/issues/1080#issuecomment-11485489), with the same caveat that you need to make sure your enums don't share any same-named values.

**Comment by andreak** [comment]
Are we really satisfied with this?
I don't think this is satisfactory at all. We use lots of Enumerations, like this:
```scala
trait ValueWithDescription[T]  {
    def description: String
    def name: String
	def wrapped: T
}

abstract class EnumWithDescriptionAndObject[T] extends Enumeration {

	abstract class ExtendedValue(id: Int) extends Val(id) with ValueWithDescription[T]

	def Value(inDescription: String, inWrapped: T) = new ExtendedValue(nextId) {
		def description = inDescription
		def name = toString()
		def wrapped = inWrapped
	}

	def Value(inWrapped: T): ExtendedValue = Value("", inWrapped)

	def getValues = {
		super.values.map(v => v.asInstanceOf[ExtendedValue]).asInstanceOf[Set[ExtendedValue]].toSeq
	}

	def valueOf(name: String) = try{Some(withName(name).asInstanceOf[ExtendedValue])} catch {case _: Throwable => None}

	def unapply(value: String) = getValues.find(_.toString == value)

}

object InvoiceStatus extends EnumWithDescriptionAndObject[(String, Int)] {
	val DRAFT = Value("Draft" -> 1)
	val IN_PROGRESS = Value("In progress" -> 2)
}
```

With the same-name restriction we won't get to serialize another Enum with value DRAFT, which we have several others of.

Enums we use *must* hold the following properties:
- Must be able to contain "extra info", hence the ```[T]```. This way we can create Enums holding arbitrary info (like i18n strings etc).
- Must be able to query the values for an Enum
- Must be able to create one from its String-representation

Other well established, Scala based, JSON frameworks use macros and/or reflection now, maybe it's time to reconsider?

**Comment by Shadowfiend** [comment]
I actually think we've been vindicated in our decision—scala.reflect will never leave experimental mode as I've understood it, as it's meant to be superseded by [scala.meta](http://scalameta.org). When 2.12 comes out it may be worth looking at building on top of scala.meta, since it seems that there are longer-term plans for its support—though time will tell.

We largely stopped using Scala Enumerations altogether in work projects a long while ago, opting instead to deal with our own as Farmer outlined above (singleton objects that extend sealed traits, with accompanying serializer). We haven't felt much if any pain from it.

That said, if you have a good serializer for the example you gave above, sharing it would be cool. I'm not morally opposed to improving the lift-json enum serializer with the implementation Franz and holograph posted, mind you, given that it's “less limited”, though still limited.

**Comment by andreak** [comment]
Scala.meta is (will be) nice, but we need solutions which work today, with today's tools, ie. scala-2.11.
I don't understand how people working with Enums don't have the 3 requirements we have.

**Comment by farmdawgnation** [comment]
What requirements are not met by the sealed trait / case object pattern above?

**Comment by farmdawgnation** [comment]
Also we're probably at the point where this discussion should be happening on the ML.

**Comment by andreak** [comment]
1. How do you hold "extra info" like a ```[T]```?
2. Create an enum-instance from its String-representation
3. Query all values of an Enum

With my abstract ```EnumWithDescriptionAndObject``` creating an Enum by extending it is easy:
```scala
object InvoiceStatus extends EnumWithDescriptionAndObject[(String, Int)] {
    val DRAFT = Value("Draft" -> 1)
    val IN_PROGRESS = Value("In progress" -> 2)
}
```
It meats the requirements:
1. Hold "extra info"
The InvoiceStatus enum holds "extra info" of type (String, Int), which is passed to Value()
1. Query all values of an Enum
InvoiceStatus.getValues
3. Create an instance form its String-representation:
InvoiceStatus.valueOf("DRAFT")

How do you do this with the sealed trait approach?

**Comment by farmdawgnation** [comment]
Ah, so I see what you mean by extra info now.

So, in all honesty you can use case classes with sealed traits as well. Those could hold more information in places where you need it. Likewise, creating a case object from it's string like representation is what the `SingletonSerializer` does. W.r.t. querying all values, you would need to maintain an `all` collection on your own, which is a little bit of boilerplate but doable. Might also be some options with Java reflection to accomplish that if you must.

**Comment by andreak** [comment]
I'm sure its doable, but when you have many Enums with those requirements then constructing them by extending ```EnumWithDescriptionAndObject``` looks *much* easier then having lots of traits/case classes.

**Comment by farmdawgnation** [comment]
I think we've reached a difference in opinion, or perhaps one in design. We've been using the trait/case object pattern for ~ 2 years now and never run into a situation where it looked difficult enough to be annoying. There's some boilerplate, but we avoided the technical debt involved in trying to figure out custom Enum support based on a library that (it looks like) will be thrown away.

There could be some discrete requirements that you have which we don't that makes that difficult for you. YMMV. But I think the right answer from the POV of the framework is to wait on meta.

Just my $0.02.

**Comment by andreak** [comment]
We're using JPA and writing converters for our kind of Enums is very easy, the converter looks like this:

```scala

// The abstract converter for our Enums
abstract class EnumConverter[T, D](et: Enumeration) extends AttributeConverter[T, String] {
	def convertToDatabaseColumn(attribute: T): String = {
		if (attribute == null) {
			null
		} else {
			attribute.toString
		}
	}

	def convertToEntityAttribute(dbData: String): T = {
		if (dbData == null) {
			null.asInstanceOf[T]
		} else {
			et.withName(dbData).asInstanceOf[T]
		}
	}

}

@Converter
class InvoiceStatusConverter extends EnumConverter[(String, Int), String](InvoiceStatus)
```
That's one line for the developer for each Enum-type. I'm not sure the converters are so easy to write using the sealed trait approach.

I think its strange that there are so many objections to provide a good solution for this serialization-problem for Scala-2.11, given that's the version 3.0 is aiming for. Isn't waiting on meta the same as saying we're aiming for 2.12? That certainly isn't the case.

Why is using Scala-reflection any worse then today's solution using java reflection and scalap/paranamer? If the argument was lack of resources to implement this I'd be OK with it but that doesn't seem to be the case reading this thread.

**Comment by farmdawgnation** [comment]
Let's move this to the ML if we want to continue the discussion where everyone can see and participate instead of having the conversation in the silo of a GitHub issue.

**Comment by Shadowfiend** [comment]
The existing code is already written and battle-tested. If a committer wants to write new code based on scala-reflect and commit to maintaining it until scala-meta, that's fine. I'm not, and Joni seems largely occupied with other things, so it doesn't look like he is either. This isn't the first time [we've had this discussion](https://groups.google.com/d/msg/liftweb/VRdhj16NKo8/21HUARyC76kJ), and the only new information that's come to light since then is the certainty that scala-reflect is going to disappear from under us, which isn't exactly a vote in its favor.

Agreed that continuing on the ML is the right way to go if we want to continue.

**Comment by andreak** [comment]
The lack of maintainer for lift-json is, IMO, a serious concern. Even if the API is stable, if it's not meeting the needs of developers developing modern apps then other tools are needed, and developers are going to look for them. I'll try to write some "not so negative" mail on the ML and move the discussion there.
